packages/firestore/src/util/bundle.ts

```diff
@@ -0,0 +1,215 @@
+  //
+  // Once a complete element is read, it is dropped from internal buffer.
+  //
+  // Returns either the bundled element, or null if we have reached the end of the stream.
+  private async nextElement(): Promise<
+    BundleElement | SizedBundleElement | null
+  > {
+    const lengthBuffer = await this.readLength();
+    if (lengthBuffer === null) {
+      return null;
+    }
+
+    const lengthString = this.textDecoder.decode(lengthBuffer);
+    const length = parseInt(lengthString, 10);
+    if (isNaN(length)) {
+      throw new Error(`length string (${lengthString}) is not valid number`);
+    }
+
+    const jsonString = await this.readJsonString(lengthBuffer.length, length);
+    // Update the internal buffer to drop the read length and json string.
+    this.buffer = this.buffer.slice(lengthBuffer.length + length);
+
+    if (!this.metadata) {
+      const element = JSON.parse(jsonString) as BundleElement;
+      return element;
+    } else {
+      return new SizedBundleElement(
+        JSON.parse(jsonString),
+        lengthBuffer.length + length
+      );
+    }
+  }
+
+  // First index of '{' from the underlying buffer.
+  private indexOfOpenBracket(): number {
+    return this.buffer.findIndex(v => v === 123);
+  }
+
+  // Reads from the beginning of the inernal buffer, until the first '{', and return
+  // the content.
+  // If reached end of the stream, returns a null.
+  private async readLength(): Promise<Uint8Array | null> {
+    let position = this.indexOfOpenBracket();
+    while (position < 0) {
+      const done = await this.pullMoreDataToBuffer();
+      if (done) {
+        if (this.buffer.length === 0) {
+          return null;
+        }
+        position = this.indexOfOpenBracket();
+        // Underlying stream is closed, and we still cannot find a '{'.
+        if (position < 0) {
+          throw new Error(
+            'Reach to the end of bundle when a length string is expected.'
+          );
+        }
+      } else {
+        position = this.indexOfOpenBracket();
+      }
+    }
+
+    return this.buffer.slice(0, position);
+  }
+
+  // Reads from a specified position from the internal buffer, for a specified
+  // number of bytes, pulling more data from the underlying stream if needed.
+  //
+  // Returns a string decoded from the read bytes.
+  private async readJsonString(start: number, length: number): Promise<string> {
+    while (this.buffer.length < start + length) {
+      const done = await this.pullMoreDataToBuffer();
+      if (done) {
+        throw new Error('Reach to the end of bundle when more is expected.');
+      }
+    }
+
+    return this.textDecoder.decode(this.buffer.slice(start, start + length));
+  }
+
+  // Pulls more data from underlying stream to internal buffer.
+  // Returns a boolean indicating whether the stream is finished.
+  private async pullMoreDataToBuffer(): Promise<boolean> {
```

Comment:
FWIW a common pattern for this type of method is to return the number of bytes read or `-1` if the end of the stream is reached.